### DISCUSSION TITLE
Protocol version 3.4 devices have a big lags sometimes

### DISCUSSION THREAD

**Discussion by rafalolb** [make-all/tuya-local#476] [General]
I have 2gang switches (from 3 different company). Switches works good with Tuya App and integrations with Tuya Cloud to HA.

All switches uses 3.4 protocal version. Switches connected to HA with this integrations works very slow. If I click switch with HA, the switch sometimes change state after 5 second. I selected 2 gang switch with timer in second form in configuration in HA. Sometimes the switches are inactive.

I need help.

Logs:
`Logger: homeassistant.helpers.entity
Source: helpers/entity.py:527
First occurred: December 30, 2022 at 15:14:34 (3873 occurrences)
Last logged: 18:17:14

Update of switch.biuroswiatla_switch_2 is taking over 10 seconds
Update of switch.salonglowne_switch_1 is taking over 10 seconds
Update of switch.salonglowne_switch_2 is taking over 10 seconds
Update of number.salonglowne_timer_1 is taking over 10 seconds
Update of number.salonglowne_timer_2 is taking over 10 seconds`

`This error originated from a custom integration.

Logger: custom_components.tuya_local.device
Source: custom_components/tuya_local/device.py:142
Integration: Tuya Local (documentation, issues)
First occurred: December 30, 2022 at 21:51:55 (839 occurrences)
Last logged: 18:17:00

Failed to refresh device state for Test.
Failed to refresh device state for SalonGlowne.
Failed to refresh device state for BiuroSwiatla.`

`Logger: homeassistant.components.number
Source: helpers/entity_platform.py:797
Integration: Number (documentation, issues)
First occurred: December 30, 2022 at 15:32:29 (1371 occurrences)
Last logged: 18:17:11

Updating tuya_local number took longer than the scheduled update interval 0:00:30`

**Comment by kadol92** [comment]
I experience the same

**Comment by Melkin1968** [comment]
The same is here on one gang switch with the protocol v3.4

**Comment by make-all** [comment]
Did you set the protocol explicitly to 3.4, or is it still on auto?  If on auto, the performance may be improved by fixing the protocol to 3.4, which you can do in the device config in the latest version.

I suspect there may still be issues with 3.4, and any errors communicating may be causing the auto protocol selection to go searching the protocol versions again, which will cause delay.

**Comment by rafalolb** [comment]
I set the protocol explicitly to 3.4

**Comment by make-all** [comment]
I'm not sure if it will help with these 3.4 device problems, but the latest change introduces persistent connections, which should improve responsiveness in general.

**Comment by rafalolb** [comment]
I described the problem on https://github.com/jasonacox/tinytuya/issues/258

I'm waiting for reply.

**Comment by make-all** [comment]
Are the lags now fixed as appears to be the feedback on the upstream report?  If so, maybe this needs to be moved to discussions, so others can find it and give ideas about what fixes these 3.4 device lags (rebooting, disabling the built-in HA tuya integration, etc)

**Comment by rafalolb** [comment]
I don't have lags now.
